Add String.toLowerCase

RobLoach opened the pull request:
I've found I've been missing a few string methods, so would like to clean it up to give Strings a bit more power. This is a small start, with `.toLowerCase`, matching [JavaScript's](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase) and [Java's](https://www.tutorialspoint.com/java/java_string_tolowercase.htm) toLowerCase() methods.

Wren's `.toLowerCase` method returns a lower-case representation of the String. This is accomplished using C's [`tolower()`](https://www.programiz.com/c-programming/library-function/ctype.h/tolower). Would like to add .toUpperCase, but wanted to get everyone's thoughts on this first.

joshgoebel commented:
There was a lot of discussion on this and proper handling of UTF-8 but if we say "we just do what C does" as a first-step I don't think that would be entirely terrible (since many other languages only deal with ASCII also).

joshgoebel commented on doc/site/modules/core/string.markdown:
```diff
@@ -249,6 +249,14 @@ Like `trim()` but only removes from the beginning of the string.
 
+### **toLowerCase**
+
+Like `trim()` but only removes from the beginning of the string.
```
This looks wrong :)

mhermier commented:
This patch is broken in many ways:
- the `const` casting to alter the content invalidate equality of the string, because the hash is created at the creation of the string.
- `tolower` is subject to locale.
- `tolower` operate `byte` array, while `String` are expected to be utf-8 encoded.
- because utf-8 is assumed there is no guarantee that the final string has the same size than the original.

mhermier commented:
@RobLoach original signature was correct and followed wren style.

RobLoach commented on src/vm/wren_core.wren.inc:
```diff
@@ -482,5 +493,4 @@ static const char* coreModuleSource =
 "  toString { \"attributes:%(_attributes) methods:%(_methods)\" }\n"
-"}\n";
-
+"}";
```
``` suggestion
"}\n";

```

RobLoach commented:
Thanks! Took on a few things...
- Moved to wren code to make use of the utf8 functional codepoints
- Renamed to `.lowerCase` to look more the other wren style functions -- Should this be `.lower` to match more Lua style?
- Made a note that it acts on only ASCII characters

mhermier commented:
While this is technically correct, because of the unfinished debate about utf-8, this should technically not be merge even with the limitation in the documentation.

PureFox48 commented:
> Should this be .lower to match more Lua style?

That would be my preference. Python also uses `lower()` but no need for us to use the parentheses.

PureFox48 commented:
@mhermier
> While this is technically correct, because of the unfinished debate about utf-8, this should technically not be merge even with the limitation in the documentation.

If you're thinking here about #948, then that was really about allowable characters in identifiers and the discussion there seemed to be drifting towards maintaining the ASCII-only status quo as most of the non-English speakers seemed comfortable with that.

I can't in any case see why we shouldn't merge this PR now as it could always be extended later if it were decided to go with some (probably table-based) approach to sort out what were lower and upper case letters within the Unicode character set as a whole.

mhermier commented:
@PureFox48 At the end #948, talks about utf-8 and what is an upper and lower case for the vm. While I agree this could go as this, it would create a fracture in the notion of string casing. Even when documented to only work with ASCII, it would create a precedence of what is lower/upper and will be relied on.

ChayimFriedman2 commented:
I don't agree about ASCII. The world is Unicode, now, and although ASCII lowercase can be useful, it is much less useful than Unicode. There is a reason why most languages use Unicode in that. If this adds too much bloat, this can be thing to leave to a library.

joshgoebel commented:
>  The world is Unicode, now,

_Even if this is conceded... I'm in favor of continuous improvement vs postponed perfection.  If we agree upcase and downcase are good things to have and we have well-tested implementations that start with ASCII-first, then lets add those as they have a lot of utility for a lot of folks.  When a well-tested PR is submitted that handles UTF-8 upcase/downcase then that can be merged at a later time.

ChayimFriedman2 commented:
Unfortunately, migrating ASCII to Unicode in this case is a breaking change. However, if we declare this from the beginning... That's probably possible.

joshgoebel commented:
> Unfortunately, migrating ASCII to Unicode in this case is a breaking change.

I'm really hoping someone will clear up what the official policy is on breaking changes.  I feel like I've only seen people offer their opinions.  To me [we can never break anything] seems a very ill-advised and harmful policy for a young project.  It prevents us from moving forward in smaller steps and leads to a "postponed perfect" vs "continuous improvement" mindset. This issue is just one example of that happening.

Perhaps we aren't using semantic versioning, but my understanding is that prior to version 1.0 things can freely break... and if there are any guarantees at all it would be only against minor releases... ie 0.4 and 0.4.1 might be compatible... 0.4 and 0.5 would not need to be.

RobLoach commented:
> Unfortunately, migrating ASCII to Unicode in this case is a breaking change.

While I can understand how you could interpret that as a breaking change, I would propose that enabling UTF-8 support to `.lower` is an added feature. This uses Wren's codepoints, which already has support for both.

joshgoebel commented:
It would be _both an added feature and a breaking change_ for anyone depending on the initial implementations "only alters ASCII character" behavior...

RobLoach commented:
Technically true, but how many people would use `.lower` and hope that it doesn't lowercase? :upside_down_face:

RobLoach commented:
Found a potential list for UTF-8 characters https://pastebin.com/fuw4Uizk

joshgoebel commented:
> Technically true, but how many people would use .lower and hope that it doesn't lowercase UTF-8? 

_With breaking changes one must be technical._  I'm of course in favor of this improvement (even as it stands).  My point was that while it most definitely is a breaking change, it's quite likely not a very *impactful* one, **and** that we're not even 1.0 yet - so we should be open to breaking changes if they push the project forward.

PureFox48 commented:
> Found a potential list for UTF-8 characters 

I think that table, which lists 1,482 letters, must be out of date as it was submitted in 2013 and Unicode has moved on quite a bit since then.

According to [Wikipedia](https://en.wikipedia.org/wiki/Unicode_character_property), as of Unicode 13.0, there were 1,791 upper case and 2,155 lower case letters so, assuming nearly all of the former have a lower case equivalent, it looks like we're up above 1,700 now.

PureFox48 commented:
I've now found the official Unicode [case charts](http://www.unicode.org/charts/case/) together with some [explanatory notes](https://unicode-org.github.io/icu/userguide/transforms/casemappings.html).

Unfortunately, the charts aren't in an ideal form for us to easily create a mapping table and there are also a few which aren't `1 to 1` which doesn't help either.

mhermier commented:
This is too much for us to maintain for now. The easiest solution is a configuration hook and a best effort fallback with ascii-7 support and no promise on the future on the best effort.

PureFox48 commented:
Well, we could certainly do ISO-8859-1(5) and we might be able to go a little further than that without using a table.

But I think trying to provide complete Unicode coverage in a small, simple language such as Wren is impractical given the complexities involved and we should instead concentrate on solutions which are relatively easy to implement and will benefit the most people.

RobLoach commented:
That's a reasonable compromise. https://cs.stanford.edu/people/miles/iso8859.html

mhermier commented:
Well the minimal requirement for the notion of casing is ascii-7 because of the capital letter naming rule. Guaranteeing more is just superfluous bonus. So even if in implementation we might do ISO-8859-1(5) I think the doc should only state about ascii-7 without extra guarantee than `it might work, or use hook to provide your casing functions`.

PureFox48 commented:
Identifiers are one thing and text handling is another.

Even if we stay with ASCII-7 for the former (because Wren uses English keywords), I don't see why text handling functions shouldn't push the boat out a little and encompass some non-English letters as well.

mhermier commented:
Identifiers already have UTF-8 requests, so identifiers are already on the same basis as text.

The thing is that by rising the minimum guarantee, you are increasing code that is required. By letting the guarantee to a minimum, if we determine that everyone use some utf-8 library via hooks, we can drop the extra support almost silently.

PureFox48 commented:
Yes, but this PR is being done using Wren code and all it would take to support the basic IS0-8859-1  character set (code-points up to 255) is this:

```javascript
lower {
    var output = ""
    for (c in codePoints) {
      if (c >= 65 && c <= 90) || (c >= 192 && c <= 214) || (c >= 216 && c <= 222)) {
        c = c + 32
      }
      output = output + String.fromCodePoint(c)
    }
    return output
}
```
So it's not as if much extra code is needed and that might be good enough for a lot of people without needing to worry about plugging in a UTF-8 case table at all.

mhermier commented:
It does not bring anything that an external function could do. And it makes precedence on the notion of casing. So it is still not good. It should call a builtin function that does the codePoint lower conversion, that the vm can hook to call an external lib i18n lib to provide complete lower casing of the value.

joshgoebel commented:
>  that the vm can hook to call an external lib i18n lib

Is there something special about string processing that warrants a new hook? (perhaps it is just that special, I'm not sure)  I just don't feel we should get in the habit of adding hooks for every little thing... it's easy enough for someone to implement `Unicode.lower` and `Unicode.upper` with the existing C API... **I'm not sure the existing String library should change it's behavior depending on a user provided hook function.**

It feels to me like we should either:

- Fully support Unicode upcase/downcase in Core
- Don't support Unicode upcase/downcase in Core, it's a user/library concern
- Realizing this is a general problem where it would be useful to extend core classes - and tackle that problem separately.

mhermier commented:
> * Fully support Unicode upcase/downcase in Core

For ease of use, this means having a reference utf-8 library. Not a big deal, but can have some issue because of licensing and conflicts with users needing a duplicate version of the same library at another version...

> * Don't support Unicode upcase/downcase in Core, it's a user/library concern

From a maintenance stand point it is basically applying functional programming. It would restrict the linking of the external library to the module, which would probably make the things more manageable.

> * Realizing this is a general problem where it would be useful to extend core classes - and tackle that problem separately.

This would solve a lot of problems, but would require a `ByteArray`/`String` or `String`/`Utf8String` couple so I don't know how impact-full it could be.

I don't know the correct answer, but I don't think doing the way it is currently done in this PR is the the way to go at all. Being simple is not an excuse/justification for the adoption in core to me. It should be either not doable by other means, or full-featured.

RobLoach commented:
> Don't support Unicode upcase/downcase in Core, it's a user/library concern
> Realizing this is a general problem where it would be useful to extend core classes - and tackle that problem separately.

My vote would be taking these two strategies. We recognize utf8 is important for Wren, but we want to keep Wren's core small and quick. 99% of the time when using upper/lower, the user would want what is provided here. Adding more would add a level of complexity that I think is best kept outside of Wren's core.

Being able to override `String`'s .lower function in a `Utf8` Wren module would be awesome, but is a limitation we don't have currently. Should likely be taken on in a new issue. In the mean time, using an external `Utf8.lower()` method would work if you need it to go beyond `String.lower`.

PureFox48 commented:
I'm not sure whether this is a good idea or not (at least as far as core is concerned)  but one way of meeting @mhermier's objection (at least in part) would be to add a second overload of `lower` which took a `caseMap` parameter:

```javascript
lower(caseMap) {
    if (!(caseMap is Map)) Fiber.abort("CaseMap must be a map")
    var output = ""
    for (c in codePoints) {
      if ((c >= 65 && c <= 90) || (c >= 192 && c <= 214) || (c >= 216 && c <= 222)) {
        c = c + 32
      } else if (caseMap.containsKey(c)) {
        c = caseMap[c]
      }
      output = output + String.fromCodePoint(c)
    }
    return output
}
```

If then, say, you wanted to increase coverage from ISO-8859-1 to ISO-8859-15, you could define your map like this:
```javascript
var caseMap = { 0x160: 0x161, 0x17d: 0x17e, 0x152: 0x153, 0x178: 0xff }
```
Anyone whose language wasn't catered for (in whole or in part) could write their own `caseMap` to plug into this method and the community at large could perhaps create maps for various languages.

Although it would be possible to combine the code for the two methods and `lower` could then call `lower(caseMap)` with a `null` map, this would slow down the former and so it's probably best to code them separately as the latter would be used much less often.

mhermier commented:
I don't understand the problem with C code like this:
```C
uint32_t wrenUtf8CodePointToLower(WrenVM* vm, uint32_t codePoint) {
  if (vm->config.utf8CodePointToLower != NULL)
  {
    return vm->config.utf8CodePointToLower(codePoint);
  }

  if ((c >=  65 && c <=  90) || // guaranteed
      (c >= 192 && c <= 214) || // optional
      (c >= 216 && c <= 222))   // optional
  {
        codePoint += 32;
  }
  return codePoint;
}
```

PureFox48 commented:
I don't think there is a problem writing this function in C (except that `c` should be `codePoint`).

If you recall, @RobLoach originally wrote it in C and then changed to Wren, apparently as a result of something you said.

mhermier commented:
The problem was that the string was produced incorrectly. So I guess he went the easy route, to avoid dealing with allocations/codepoints. As far as I know, no guarantee that lower/upper case code point takes the same amounts of bytes, so it mess a little bit more allocation of a temporary buffer, before `String` creation.

PureFox48 commented:
If we restrict it to code-points less than 255, then we an guarantee that the upper/lower case characters will take the same number of bytes but beyond that there will certainly be exceptions.

joshgoebel commented:
Yes, I felt like we were slightly moving toward "ASCII is enough" for now, and if that is a given then the byte count is always the same just a transformation...
